server/connmux/connmux.go

```diff
@@ -0,0 +1,332 @@
+		c.lg.Debug("forwarding connection to the HTTP backend", zap.String("remote address", conn.RemoteAddr().String()))
+		select {
+		case c.http.connc <- conn:
+		case <-c.donec:
+		}
+		return
+	}
+
+	if c.grpc != nil {
+		c.lg.Debug("forwarding connection to the GRPC backend", zap.String("remote address", conn.RemoteAddr().String()))
+		select {
+		case c.grpc.connc <- conn:
+		case <-c.donec:
+		}
+		return
+	}
+
+	log.Println("unknown connection")
+	conn.Close()
+}
+
+// Close closes the listeners
+func (c *ConnMux) Close() error {
+	time.Sleep(gracefullShutdownDuration)
+	c.closeDoneChans()
+	return c.root.Close()
+}
+
+func (c *ConnMux) closeDoneChans() {
+	select {
+	case <-c.donec:
+	default:
+		close(c.donec)
+	}
+}
+
+// muxListener is the listener exposed to the HTTP and GRPC servers
+// The root listener Accept() method is overriden
+// The multiplexed servers have access to the Close() and Address()
+// methods on the root listener.
+type muxListener struct {
+	net.Listener
+	connc chan net.Conn
+	donec chan struct{}
+}
+
+var _ net.Listener = (*muxListener)(nil)
+
+func (l muxListener) Accept() (net.Conn, error) {
+	select {
+	case c, ok := <-l.connc:
+		if !ok {
+			return nil, net.ErrClosed
+		}
+		return c, nil
+	case <-l.donec:
+		return nil, net.ErrClosed
+	}
+}
+
+// HTTPListener returns a net.Listener that will receive http requests
+func (c *ConnMux) HTTPListener() net.Listener {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.http == nil {
+		c.http = &muxListener{c.root, make(chan net.Conn), c.donec}
+	}
+	return c.http
+}
+
+// GRPCListener returns a net.Listener that will receive grpc requests
+func (c *ConnMux) GRPCListener() net.Listener {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.grpc == nil {
+		c.grpc = &muxListener{c.root, make(chan net.Conn), c.donec}
+	}
+	return c.grpc
+}
+
+// bufferedConn allows to peek in the buffer of the connection
+// without advancing the reader.
+type bufferedConn struct {
+	r *bufio.Reader
+	net.Conn
+}
+
+func newBufferedConn(c net.Conn) bufferedConn {
+	return bufferedConn{bufio.NewReader(c), c}
+}
+
+func (b bufferedConn) Peek(n int) ([]byte, error) {
+	return b.r.Peek(n)
+}
+
+func (b bufferedConn) Read(p []byte) (int, error) {
+	return b.r.Read(p)
+}
+
+func isHTTP2Connection(r io.Reader) bool {
+	// check the http2 client preface
+	buf := make([]byte, len(http2.ClientPreface))
+	n, err := r.Read(buf)
+	if err != nil || n != len(http2.ClientPreface) {
+		return false
+	}
+	return bytes.Equal(buf, []byte(http2.ClientPreface))
+}
+
+func isGRPCConnection(lg *zap.Logger, r io.Reader) bool {
+	done := false
+	isGRPC := false
+	// check if len of the settings frame is 0 or the headers with the "content-type"
+	// indicates is an grpc connection.
+	framer := http2.NewFramer(io.Discard, r)
+	// match headers names or values based on https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+	hdec := hpack.NewDecoder(uint32(4<<10), func(hf hpack.HeaderField) {
```

Comment:
do you mind leaving a comment here why the decoder needs a `4<<10` shift? I wasn't able to find this in the markdown you've linked.